viewing_party/party.py

```diff
@@ -1,23 +1,180 @@
+    for i in range(len(user_data['friends'])):
+        for movie in user_data['friends'][i]['watched']:
+            friends_movies.append(movie['title'])
 
+    #convert to set to get unique titles via difference method
+    user_watched_set = set(user_watched_titles)
+    friends_movies_set = set(friends_movies)
+    unique_watched = user_watched_set - friends_movies_set
+    
+    #generate list of unique movie dictionaries
+    for movie in user_data['watched']:
+        if movie['title'] in unique_watched:
+            unique_movies.append(movie)
+
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+
+    user_watched_titles = []
+    friends_movies = []
+    friends_unique_movies = []
+    
+    for movie in user_data['watched']:
+        user_watched_titles.append(movie['title'])
+    
+    for i in range(len(user_data['friends'])):
+        for movie in user_data['friends'][i]['watched']:
+            friends_movies.append(movie['title'])
+
+    user_watched_set = set(user_watched_titles)
+    friends_movies_set = set(friends_movies)
+    unique_watched = friends_movies_set - user_watched_set
+    
+    #avoiding duplicate movies by not adding movie that 
+    #is already in friends_unique_movies list
+
+    for i in range(len(user_data['friends'])):
+        for movie in user_data['friends'][i]['watched']:
+            if movie['title'] in unique_watched and movie not in friends_unique_movies:
+                 friends_unique_movies.append(movie)
+    
+    return friends_unique_movies
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    movie_recs = []
+    friends_unique_watched = get_friends_unique_watched(user_data)
+    
+    for movie in friends_unique_watched:
+        if movie['host'] in user_data['subscriptions']:
+            movie_recs.append(movie)
+
+    return movie_recs
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    movie_recs = []
+    user_watched = []
+    popular_genre = get_most_watched_genre(user_data)
+    friends_unique_watched = get_friends_unique_watched(user_data)
+    
+    for movie in user_data['watched']:
+        user_watched.append(movie['title'])
+
+    for movie in friends_unique_watched:
+        if movie['title'] not in user_watched and movie['genre'] == popular_genre:
+            movie_recs.append(movie)
+
+    return movie_recs
+
+def get_rec_from_favorites(user_data):
+    movie_recs = []
+    user_unique_movies = get_unique_watched(user_data)
+    user_unique_movie_titles = []
+
+    for movie in user_unique_movies:
+        user_unique_movie_titles.append(movie['title'])
+
+    for movie in user_data['favorites']:
+        if movie['title'] in user_unique_movie_titles:
+            movie_recs.append(movie)

```

Comment:
We can also shorten this two for-loop logic into one for loop and get rid of the intermediate `user_unique_movie_titles` list, if desired:
```
for movie in user_unique_movies:
    if movie in user_data['favorites']:
        movie_recs.append(movie)
```